### PR TITLE
Issue #8

### PR DESCRIPTION
**Pull request by TobiUUlm**
Fixes https://github.com/Universal-Variability-Language/uvl-lsp/issues/8  

### How it works:  
Open a `.uvl` file, run the `Generate a DOT file to visualize with Graphviz` command (Open Command Palette with `Ctrl + Shift + P`).  
The LSP will create a `<sameFileName>.dot` file (in the same directory) and visualize its content via [this Graphviz Preview Extension](https://marketplace.visualstudio.com/items?itemName=tintinweb.graphviz-interactive-preview).  
Any changes in this file (as well as repeatedly generating it) will produce a live-preview of the graph (while editing).  

issue_8

The [VSCode Extension (for Graphviz preview)](https://marketplace.visualstudio.com/items?itemName=tintinweb.graphviz-interactive-preview) is marked as a  extensionDependency and will therefore be loaded with the UVLS Extension. Please evaluate if this has any effect on privacy policy or needs to be mentioned beforehand.

To test this feature, you need to use `cargo build` to get the updated server as well as run the extension.  

If this PR is accepted, we probably need to run GitHub actions to build a new release (both for the server and extension), if we want this to be included immediately.  
However, I don't have any experience in this field.

### What might be added:
We can think about a lot of ways to improve this:  
- Add a button to simply create a visualization
- Add a keyboard-shortcut to create a visualization
- Add a Legend (currently just commented-out)
- Add another Visualization tool (quite easy to switch)
- Show cardinality of Features/Groups
- Add more information as a tooltip
- Add configuration support

### CODE REVIEW

**Comment by ThiBruUU** [comment]
On Windows there are no problems with the installation and the generation of the dot file works fine
LGTM

**Review by ThiBruUU — APPROVED** [review]

**Comment by TobiUUlm** [comment]
Now it should work on Windows and Non-Windows machines.  
Please test/review again @SundermannC @pascalfoerster @felixrieg @ThiBruUU

File: uvls/src/main.rs [hunk 1]
```diff
@@ -389,6 +392,27 @@ impl LanguageServer for Backend {
...
+                let g = ast::graph::Graph::new(document.source.clone(), document.tree.clone(), uri.clone());
+
+                // write graph script (dot) to file:
+                let diagram_file_extension = "dot";
+                let re = regex::Regex::new(r"(.*\.)(.*)").unwrap();
+                let path = re.replace(uri.path(), |caps: &regex::Captures| {format!("{}{}", &caps[1], diagram_file_extension)});
+                let file = std::fs::File::create(path.as_ref())
+                    .or(std::fs::File::create(percent_decode_str(&path.replacen("/", "", 1)).decode_utf8().unwrap().into_owned())); // windows specific
+
+                if !file.is_err() {
```
Comment (TobiUUlm) [inline uvls/src/main.rs:410]: change to `is_ok` ?

File: editors/vscode/README.md [hunk 2]
```diff
@@ -19,7 +19,9 @@ Currently supports
 On most common platforms installing this extension should just work, without any dependencies.
 When the automatic install fails you will need to build the server you're self.
 Build instructions can be found [here](https://codeberg.org/caradhras/uvls).
-To use a custom build server specify its location via `uvls.path` or add it to `PATH`
+To use a custom build server specify its location via `uvls.path` or add it to `PATH`  
+To use the Visualization of Feature Models please install the [Graphviz Interactive Preview Extension](https://marketplace.visualstudio.com/items?itemName=tintinweb.graphviz-interactive-preview) (tintinweb.graphviz-interactive-preview).
+You may use any other Graphviz Preview Extension, but this highly compatible with the UVLS Extension.
```
Comment (TobiUUlm) [inline editors/vscode/README.md:24]: "... but this **is** highly compatible with **our** UVLS Extension."

File: uvls/src/main.rs [hunk 3]
```diff
@@ -389,6 +392,27 @@ impl LanguageServer for Backend {
...
+                if !root_graph.contains_id(root_fileid){{}}
+                
+                let document = root_graph.files.get(&root_fileid).unwrap();
+                let g = ast::graph::Graph::new(document.source.clone(), document.tree.clone(), uri.clone());
+
+                // write graph script (dot) to file:
+                let diagram_file_extension = "dot";
+                let re = regex::Regex::new(r"(.*\.)(.*)").unwrap();
+                let path = re.replace(uri.path(), |caps: &regex::Captures| {format!("{}{}", &caps[1], diagram_file_extension)});
+                let file = std::fs::File::create(path.as_ref())
```
Comment (TobiUUlm) [inline uvls/src/main.rs:407]: We might also think about not creating an additional file at all.  
Or maybe adding a setting for this.  
Creating a file might be useful for editing the graph later on.

**Comment by SundermannC** [comment]
Thank you for the changes! The visualization now works great for me on Ubuntu.

File: uvls/src/main.rs (hunk 3 above)
Comment (SundermannC) [reply uvls/src/main.rs:407]: How would this work without creating a .dot file? Or what do you mean?

Comment (TobiUUlm) [reply uvls/src/main.rs:407]: We can simply return the Graph-Script (as we already do as soon as the file creation fails) and pass it via our vsCode extension to the GraphViz Extension that is able to display it without having to read a `.dot` file.  
However, this prevents the end-user to edit the Graph-Script in any custom way. That's why I think this should only be a fallback solution.

**Comment by TobiUUlm** [comment]
Ready to merge :) @SundermannC

**Review by pascalfoerster — APPROVED** [review]
LGTM

**Comment by ThiBruUU** [comment]
to generate the graph, an additional button was added next to the configure button

**Comment by SundermannC** [comment]
The new button is a great addition. For me, the button however only generates the .dot file, while the command also opens the view. I do not really mind, but is this intended? @ThiBruUU

**Comment by ThiBruUU** [comment]
currently it is not yet implemented that it opens directly. But that can still be submitted later.

**Comment by TobiUUlm** [comment]
We added explicit vsCode button and keybinding support (Ctrl + G) to generate a graph.  
This triggers the vsCode command and therefore works just as running the original command.
You can see how this works, here:
issue_8_button

This probably needs a separate PR.

**Comment by TobiUUlm** [comment]
[New PR here](https://github.com/Universal-Variability-Language/uvl-lsp/pull/33).

**Pull request merged**